goblin/driver/connection.py

```diff
@@ -35,17 +32,14 @@ def __init__(self, response_queue, loop, conn, username, password,
             raise StopAsyncIteration
 
     async def fetch_data(self):
-        if not self._response_queue.empty():
-            message = self._response_queue.get_nowait()
-        else:
-            self._loop.create_task(self._conn.get_data())
-            message = await self._response_queue.get()
-        return message
```

Comment:
Here was the main issue. First of all, we're waiting to read from the socket until the user reads from the response. Meanwhile the incoming data on the socket is building up in memory, presumably in an internal structure in the aiohttp websocket client.

Also it makes sense to await getting from the queue. Maybe the next message isn't available just yet, so instead of triggering an exception we just await.

Also, the task for the connection to receive another message should be triggered under exactly two conditions: the most recently retrieved message was a 206, or the most recently retrieved message was a 407 and we just sent auth credentials. That's now handled in the connection.

 In the new version, the response doesn't need to know anything about its connection.